### PR TITLE
init: Fix asmap/addrman initialization order bug

### PR DESCRIPTION
**Pull request by jnewbery**
Commit 181a1207 introduced an initialization order bug: CAddrMan's m_asmap must be set before deserializing peers.dat.

The first commit restores the correct initialization order. The remaining commits make `CAddrMan::m_asmap` usage safer:

- don't reach into `CAddrMan`'s internal data from `CConnMan`
- set `m_asmap` in the initializer list and make it const
- make `m_asmap` private, and access it (as a reference to const) from a getter.

This ensures that peers.dat deserialization must happen after setting m_asmap, since m_asmap is set during CAddrMan construction.

### CODE REVIEW

**Comment by jonatack** [comment]
I'm unable to build (with clang 13, CI seems to agree):
```
bench/addrman.cpp:76:18: error: no matching constructor for initialization of 'CAddrMan'
        CAddrMan addrman{/* deterministic */ false, /* consistency_check_ratio */ 0};
                 ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./addrman.h:181:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
class CAddrMan
      ^
./addrman.h:458:14: note: candidate constructor not viable: requires 3 arguments, but 2 were provided
    explicit CAddrMan(std::vector<bool>* asmap, bool deterministic, int32_t consistency_check_ratio);
             ^
bench/addrman.cpp:83:14: error: no matching constructor for initialization of 'CAddrMan'
    CAddrMan addrman(/* deterministic */ false, /* consistency_check_ratio */ 0);
             ^                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./addrman.h:181:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
class CAddrMan
      ^
./addrman.h:458:14: note: candidate constructor not viable: requires 3 arguments, but 2 were provided
    explicit CAddrMan(std::vector<bool>* asmap, bool deterministic, int32_t consistency_check_ratio);
             ^
bench/addrman.cpp:95:14: error: no matching constructor for initialization of 'CAddrMan'
    CAddrMan addrman(/* deterministic */ false, /* consistency_check_ratio */ 0);
             ^                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./addrman.h:181:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
class CAddrMan
      ^
./addrman.h:458:14: note: candidate constructor not viable: requires 3 arguments, but 2 were provided
    explicit CAddrMan(std::vector<bool>* asmap, bool deterministic, int32_t consistency_check_ratio);
             ^
```

**Comment by jonatack** [comment]
Testing the first commit.

**Comment by jnewbery** [comment]
> I'm unable to build (with clang 13, CI seems to agree)

Oops. I don't build with bench locally to save a bit of time. Now fixed.

**Comment by jonatack** [comment]
First commit: the addrman checks pass, but I lose 18k addresses on startup and then see parsing errors with the asmap file. 
```
2021-08-24T13:39:14Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
2021-08-24T13:39:15Z [init] Using asmap version 7ead927689f884bc4021618017cc1438d9328578779edd2adfbed044fb52928e for IP bucketing
2021-08-24T13:39:15Z [init] init message: Loading P2P addresses…
2021-08-24T13:39:16Z [init] Bucketing method was updated, re-bucketing addrman entries from disk
2021-08-24T13:39:25Z [init] addrman lost 18599 new and 1 tried addresses due to collisions or invalid addresses
2021-08-24T13:39:25Z [init] Addrman checks started: new 26373, tried 72, total 26445
2021-08-24T13:39:25Z [init] Addrman checks completed successfully
2021-08-24T13:39:25Z [init] Loaded 26445 addresses from peers.dat  9577ms
...
Error: Error reading configuration file: parse error on line 334: 2021-08-24T13:39:14Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
```

**Comment by jonatack** [comment]
Test with only the first commit: the addrman checks pass throughout, but I lost 18k addresses on startup.
```
2021-08-24T13:39:14Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
2021-08-24T13:39:15Z [init] Using asmap version 7ead927689f884bc4021618017cc1438d9328578779edd2adfbed044fb52928e for IP bucketing
2021-08-24T13:39:15Z [init] init message: Loading P2P addresses…
2021-08-24T13:39:16Z [init] Bucketing method was updated, re-bucketing addrman entries from disk
2021-08-24T13:39:25Z [init] addrman lost 18599 new and 1 tried addresses due to collisions or invalid addresses
2021-08-24T13:39:25Z [init] Addrman checks started: new 26373, tried 72, total 26445
2021-08-24T13:39:25Z [init] Addrman checks completed successfully
2021-08-24T13:39:25Z [init] Loaded 26445 addresses from peers.dat  9577ms
```

Update: deleted my peers.dat and re-tried building on the PR branch after the update; debug build now succeeds and retesting with multiple restarts after deleting peers.dat initially seems fine.
```
2021-08-24T14:11:09Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
2021-08-24T14:11:11Z [init] Using asmap version 7ead927689f884bc4021618017cc1438d9328578779edd2adfbed044fb52928e for IP bucketing
2021-08-24T14:11:11Z [init] init message: Loading P2P addresses…
2021-08-24T14:11:11Z [init] Addrman checks started: new 6828, tried 10, total 6838
2021-08-24T14:11:12Z [init] Addrman checks completed successfully
2021-08-24T14:11:12Z [init] Loaded 6838 addresses from peers.dat  756ms
```

<details><summary>tested with these config args</summary><p>

```
2021-08-24T13:41:49Z [init] Config file arg: blockfilterindex="1"
2021-08-24T13:41:49Z [init] Config file arg: checkaddrman="1"
2021-08-24T13:41:49Z [init] Config file arg: checkblocks="12"
2021-08-24T13:41:49Z [init] Config file arg: checklevel="4"
2021-08-24T13:41:49Z [init] Config file arg: checkmempool="1"
2021-08-24T13:41:49Z [init] Config file arg: coinstatsindex="1"
2021-08-24T13:41:49Z [init] Config file arg: debug="tor"
2021-08-24T13:41:49Z [init] Config file arg: debug="i2p"
2021-08-24T13:41:49Z [init] Config file arg: debug="addrman"
2021-08-24T13:41:49Z [init] Config file arg: i2psam="127.0.0.1:7656"
2021-08-24T13:41:49Z [init] Config file arg: logips="1"
2021-08-24T13:41:49Z [init] Config file arg: logthreadnames="1"
2021-08-24T13:41:49Z [init] Config file arg: par="1"
2021-08-24T13:41:49Z [init] Config file arg: peerbloomfilters="1"
2021-08-24T13:41:49Z [init] Config file arg: persistmempool="1"
2021-08-24T13:41:49Z [init] Config file arg: txindex="1"
2021-08-24T13:41:49Z [init] Config file arg: walletbroadcast="1"
2021-08-24T13:41:49Z [init] Config file arg: [main] asmap="/home/jon/projects/bitcoin/asmap/demo.map"
```
</p></details>

**Comment by jonatack** [comment]
<details><summary>Further testing seems ok</summary><p>

Restarted after turning off -asmap:

```
2021-08-24T14:19:00Z [init] Using /16 prefix for IP bucketing
2021-08-24T14:19:00Z [init] init message: Loading P2P addresses…
2021-08-24T14:19:01Z [init] Bucketing method was updated, re-bucketing addrman entries from disk
2021-08-24T14:19:01Z [init] addrman lost 2147 new and 0 tried addresses due to collisions or invalid addresses
2021-08-24T14:19:01Z [init] Loaded 9928 addresses from peers.dat  463ms
```
Restarted again with -asmap off:
```
2021-08-24T14:20:59Z [init] Using /16 prefix for IP bucketing
2021-08-24T14:20:59Z [init] init message: Loading P2P addresses…
2021-08-24T14:21:00Z [init] Loaded 10760 addresses from peers.dat  169ms
```
Restarted with -asmap back on:
```
2021-08-24T14:22:44Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
2021-08-24T14:22:46Z [init] Using asmap version 7ead927689f884bc4021618017cc1438d9328578779edd2adfbed044fb52928e for IP bucketing
2021-08-24T14:22:46Z [init] init message: Loading P2P addresses…
2021-08-24T14:22:46Z [init] Bucketing method was updated, re-bucketing addrman entries from disk
2021-08-24T14:22:48Z [init] addrman lost 510 new and 0 tried addresses due to collisions or invalid addresses
2021-08-24T14:22:48Z [init] Loaded 10462 addresses from peers.dat  2802ms
```
Restarted again with -asmap on and reduced frequency of checkaddrman=100, leaving running this way for now:
```
2021-08-24T14:24:41Z [init] Opened asmap file "/home/jon/projects/bitcoin/asmap/demo.map" (932999 bytes) from disk
2021-08-24T14:24:43Z [init] Using asmap version 7ead927689f884bc4021618017cc1438d9328578779edd2adfbed044fb52928e for IP bucketing
2021-08-24T14:24:43Z [init] init message: Loading P2P addresses…
2021-08-24T14:24:43Z [init] Loaded 11203 addresses from peers.dat  737ms
...
2021-08-24T14:30:12Z [msghand] Addrman checks started: new 15479, tried 19, total 15498
2021-08-24T14:30:13Z [msghand] Addrman checks completed successfully
```

Addrman checks passing throughout.

</p></details>

File: src/init.cpp
Comment (vasild) [inline src/init.cpp:1179]: What is the point of `{}`?

File: src/addrman.h
Comment (jonatack) [inline src/addrman.h:580]: 4ffc8b81bc520fe66fb is there any benefit to making `m_asmap` private rather than a public data member? referencing our previous conversation:

- https://github.com/bitcoin/bitcoin/pull/20210#issuecomment-766093925

- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c131-avoid-trivial-getters-and-setters

(otherwise, make the getter const?)
```suggestion
    const std::vector<bool>& GetAsmap() const { return m_asmap; }
```

**Review by vasild — COMMENTED** [review]
ACK 4ffc8b81bc520fe66fbea9a597ab889a2647ca2e

File: src/init.cpp
Comment (jonatack) [inline src/init.cpp:1179]: 47ad0fd2644c4de7a11efa9c9e288f73d6c147e5 naming nit: `asmap_data` would allow more easily grepping for / distinguishing this new data structure in the code

File: src/init.cpp
Comment (jonatack) [inline src/init.cpp:1218]: 47ad0fd2644c4de7a11efa9c9e288f73d6c147e5 nit, why these outer brackets?

**Review by jonatack — COMMENTED** [review]
Concept ACK 4ffc8b81bc520fe66fbea9a597ab889a2647ca2e modulo comments below and feedback from fuzzing people wrt invalidating affected corpii

Suggestions:

- link to https://www.erisian.com.au/bitcoin-core-dev/log-2021-08-23.html#l-263 (up to line 382) in the PR description for context
- describe the symptoms/consequences of the issue and how to reproduce
- maybe save refactoring/renaming for a non-bug-fix
- add a passing/failing test that demonstrates the bug is fixed

File: src/init.cpp
Comment (jnewbery) [reply src/init.cpp:1179]: Removed

File: src/addrman.h
Comment (jnewbery) [reply src/addrman.h:580]: Thanks. Done.

File: src/test/fuzz/addrman.cpp
Comment (jnewbery) [reply src/test/fuzz/addrman.cpp:313]: Done

File: src/addrman.h
Comment (jnewbery) [reply src/addrman.h:580]: I've made the getter const. I think removing the only public data member of CAddrMan is an improvement.

File: src/init.cpp
Comment (jnewbery) [reply src/init.cpp:1179]: Good suggestion, but will hold off doing it in this PR to reduce the diff. This part is mostly move-only.

File: src/init.cpp
Comment (jnewbery) [reply src/init.cpp:1218]: It limits the scope and makes it obvious that local variables like `asmap` and `adb` are not used outside this code block.

**Comment by jnewbery** [comment]
Thanks for the reviews @vasild @jonatack. I've addressed all your inline comments.

@jonatack - If you have any suggested wording, I'm happy to incorporate it into the PR description.

**Review by vasild — APPROVED** [review]
ACK d4fda4f9b9c9e0cebf16381f6f49b39a43ef9ef4

**Review by vasild — APPROVED** [review]
ACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7

File: src/addrman.cpp
Comment (mzumsande) [reply src/addrman.cpp:84]: Tested ACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7

Code changes look correct to me.
I verified that current master failed the addrman checks (`-checkaddrman`) when switching from no-asmap mode to asmap mode with an existing `peers.dat`, while I experienced no problems over multiple switches on this branch.

**Comment by mzumsande** [comment]
Tested ACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7

Code changes look correct to me.
I verified that current master failed the addrman checks (`-checkaddrman`) when switching from no-asmap mode to asmap mode with an existing `peers.dat`, while I experienced no problems over multiple switches on this branch.

**Comment by jonatack** [comment]
> Suggestions:
> 
>     * link to [erisian.com.au/bitcoin-core-dev/log-2021-08-23.html#l-263](https://www.erisian.com.au/bitcoin-core-dev/log-2021-08-23.html#l-263) (up to line 382) in the PR description for context
> 
>     * describe the symptoms/consequences of the issue and how to reproduce
> 
>     * maybe save refactoring/renaming for a non-bug-fix
> 
>     * add a passing/failing test that demonstrates the bug is fixed

Addressed these suggestions in #22831 that pulls in the first commit here. That patch could be merged first and this one rebased on it (or vice-versa), or you could optionally pull in the test commits and use the PR description--as you prefer.

File: src/addrman.cpp
Comment (amitiuttarwar) [inline src/addrman.cpp]: question about use of `std::move` here - 

If reading `peers.dat` causes issues, `init.cpp` will instantiate a `CAddrMan` twice, using the same asmap both times. Is this ok? I know that `std::move` is more of a suggestion than a command, so I'm wondering how this code would be interpreted.

File: src/addrman.cpp
Comment (amitiuttarwar) [inline src/addrman.cpp:84]: commit: 5840476714ffebb2599999c85a23b52ebcff6090  _[addrman] Make m_asmap private_

Is making the ordering of private members last in the initializer list a personal preference? I find it slightly odd that the initializer list no longer matches the order of the ctor args. Also you could have just introduced them in this order? =P

**Comment by amitiuttarwar** [comment]
approach ACK, I want to understand the `std::move` implications & then I'm ready to leave a review ACK. 
did you consider making `m_asmap` an optional arg for the addrman constructor? 
but overall, I like these changes- in addition to fixing the bug, making asmap a private & read-only member of addrman seems like an incremental improvement to me.

File: src/addrman.cpp
Comment (vasild) [reply src/addrman.cpp]: The variable from `init.cpp` will be copied to the `CAddrMan` constructor into a temporary, used for the argument (this temporary is what is available inside the body of the constructor and will be automatically destroyed when the constructor returns). Then the constructor moves that temporary into the `m_asmap` member. So, the caller's variable from `init.cpp` will not be modified and can be reused.

See also https://github.com/bitcoin/bitcoin/pull/22791#discussion_r697472059

File: src/addrman.cpp
Comment (vasild) [reply src/addrman.cpp:84]: The order here must be the same as the order in which the variables are declared in the class, otherwise we get a compiler warning.

Comment (amitiuttarwar) [reply src/addrman.cpp:84]: ah right, thanks

File: src/addrman.cpp
Comment (amitiuttarwar) [reply src/addrman.cpp]: hm interesting, I've made some little toy programs to try to understand this behavior, and I see that the caller's variable can be reused, so that resolves my concern.

I'm not seeing the behavior of copying the temporary, but I'm going to resolve this conversation and explain further on the comment you linked.

**Comment by amitiuttarwar** [comment]
code review but utACK 724c497562

**Comment by naumenkogs** [comment]
utACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7

File: src/addrman.cpp
Comment (MarcoFalke) [reply src/addrman.cpp:84]: The order could have been kept the same if it was moved to before `insecure_rand` in the header file

File: src/addrman.h
Comment (MarcoFalke) [inline src/addrman.h:580]: please add LIFETIMEBOUND when returning a pointer to memory in `this`

**Review by maflcko — APPROVED** [review]
review ACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7 👫

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 724c4975622bc22cedc3f3814dfc8e66cf8371f7 👫
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjbBgwAmjfKpl2pPRl+MlWki+w1+qRbTqdMIiBrEu6wt79pkznSY6fNvWXAXCiI
xyqHDMf8psDsZiL3+gPq7qW5cBBPTZb8HxYvJAIqo2J5Gz2aqzNrgB4vegom0NgH
/QTVsdsgIssn2YmMXE6LpWU2nYaMgPqJPKAu4D3GZ2IEDitSLYYNGcm4M9MDYDQ3
FP+QNf+7qTE29UshNXP86iSQ2ITuKN2eXuA+Zm6BpaFmG8lkIsfdjJFQVdDDbCWX
7lB1M/v5mrIK9LylIv/X6iYJi2FMT4UdqICoudp4L+V1m7lElw3Z/ymMjaPuM0NV
9o+sJUbL+/lTFH3yncplmwkEZDdv+M4q00csIw5wAYvgpTRw0aP6YJPL1VzUISsK
R7CStBUNuIlX42hnGwHqUhs4fxwzevfg1tRTJDkk3+Z/y/E3u+Hj+eEBfoDBnU5C
/eqAT4WXAhdVZdqUcxR841Aki67Agsb2bv+pv9sAskG5OTHyynccu3T4Eh4XwsWL
MS67bp/6
=zBMj
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9a48951e3c61060be14eb894092e9ac4ae9406fa411fed5ca58fea1d1415078b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089a48951e3c61060be14eb894092e9ac4ae9406fa411fed5ca58fea1d1415078bf010d38c23cf8a8b424a22c048bdb36faa0308fff0102f87fc1b634b0eb64cdb5a921328d7eb08f1046135f038f0083c6bfce9f9897dd00083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01056573c652a337d3b53a70c65a28a65b708f1046135f038f008e60c69dbfb5e15cb0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01064b855e61223e3b3cfbff4db386c8e7d08f12017932f23404335987cbc786c5aae5d28ea56945907cb7528ddb60edf180e352208f1046135f037f00880f296ab37a37f960083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f37516dacf46c7441f503804bee41c3908f1046135f038f008ba4c6797bd3d83a90083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/addrman.h
Comment (jnewbery) [reply src/addrman.h:580]: Good idea.

I actually plan to fully encapsulate asmap in its own component in #22910, so we're not passing out references to private data.

**Comment by jamesob** [comment]
> Addressed these suggestions in #22831 that pulls in the first commit here. That patch could be merged first and this one rebased on it (or vice-versa), or you could optionally pull in the test commits and use the PR description--as you prefer.

Why was this comment, made 12 days ago and 8 days before merge, ignored? 

It would be one thing if this change was a straightforward revert or a simple reordering, but at face value it is not obvious that this change fixes the fairly severe bug that @jonatack discovered. It is troubling that a patch of this complexity was merged without an accompanying test verifying its efficacy, especially when one was made readily available.

**Comment by jonatack** [comment]
> Concept ACK [4ffc8b8](https://github.com/bitcoin/bitcoin/commit/4ffc8b81bc520fe66fbea9a597ab889a2647ca2e) modulo comments below and feedback from fuzzing people wrt invalidating affected corpii
> 
> Suggestions:
> 
>     * link to [erisian.com.au/bitcoin-core-dev/log-2021-08-23.html#l-263](https://www.erisian.com.au/bitcoin-core-dev/log-2021-08-23.html#l-263) (up to line 382) in the PR description for context
> 
>     * describe the symptoms/consequences of the issue and how to reproduce
> 
>     * maybe save refactoring/renaming for a non-bug-fix
> 
>     * add a passing/failing test that demonstrates the bug is fixed

@jamesob I didn't understand either. The suggestions in https://github.com/bitcoin/bitcoin/pull/22791#pullrequestreview-737430387 were also essentially ignored.

**Comment by MarcoFalke** [comment]
There were 5 ACKs, including one tested ACK, before I added the 6th ACK. I think this number of review ACKs (in addition to the tests to verify the fix by people who didn't leave a comment) is sufficient to merge a pull request. This is overall a higher than average review interest compared to other pull reqeusts. If the alternative pull doesn't get any ACKs, it would appear slightly odd to me to merge it instead.

**Comment by jnewbery** [comment]
> Why was this comment, made 12 days ago and 8 days before merge, ignored?
> 
> It would be one thing if this change was a straightforward revert or a simple reordering, but at face value it is not obvious that this change fixes the fairly severe bug that @jonatack discovered. It is troubling that a patch of this complexity was merged without an accompanying test verifying its efficacy, especially when one was made readily available.

@jamesob (https://github.com/bitcoin/bitcoin/pull/22791#issuecomment-917050278)

> The suggestions in #22791 (review) were also essentially ignored.

@jonatack (https://github.com/bitcoin/bitcoin/pull/22791#issuecomment-917108864)

Please see https://github.com/bitcoin/bitcoin/pull/22831#issuecomment-918092775.

**Pull request merged**
